letsencrypt/tests/auth_handler_test.py

```diff
@@ -148,6 +152,66 @@ def test_perform_failure(self):
         self.assertRaises(
             errors.AuthorizationError, self.handler.get_authorizations, ["0"])
 
+    @mock.patch("acme.challenges.HTTP01Response.simple_verify")
+    @mock.patch("acme.challenges.TLSSNI01Response.simple_verify")
+    @mock.patch("letsencrypt.auth_handler.logger")
+    def test_simple_verify(self, mock_logger, mock_tlssni01_verify, mock_http01_verify):
+        account_key = mock.Mock()
+        challenge_http_01 = achallenges.KeyAuthorizationAnnotatedChallenge(
+            challb=acme_util.HTTP01_P, domain=b'localhost', account_key=account_key)
+        challenge_tls_sni_01 = achallenges.KeyAuthorizationAnnotatedChallenge(
+            challb=acme_util.TLSSNI01_P, domain=b'localhost', account_key=account_key)
+
+        response_http_01 = challenges.HTTP01Response(key_authorization=u'foo')
+        response_tls_sni_01 = challenges.TLSSNI01Response(key_authorization=u'foo')
+
+        # Test _simple_verify's behaviour when implementation of challenge is not implemented
+        response_dummy = mock.Mock()
+        challenge_dummy = mock.Mock(chall=mock.Mock(), uri="foo", typ="Mock")
+
+        self.handler.dv_c = [challenge_http_01,
+                             challenge_tls_sni_01,
+                             challenge_dummy]
+
+        self.mock_dv_auth.perform.side_effect = [[response_http_01,
+                                                  response_tls_sni_01,
+                                                  response_dummy]]
+
+        mock_http01_verify.return_value = False
+        mock_tlssni01_verify.return_value = False
```

Comment:
Ah, you're completely right regarding the fact that we don't test the return value of ``_simple_verify`` (and yes I meant ``_simple_verify`` above as well). I'll work on that.

I understand the other problems you've mentioned but I'm not sure what would be the best way to fix such problems. I really appreciate the code snippets, though, really helped to resolve the confusion :)